File: Leetcode/Python/_290.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        hash_map = {}
        reverse_map = {}
        splited = s.split(" ")
        if len(pattern) != len(splited):
            return False
        for pat, match in zip(pattern, splited):
            if pat not in hash_map:
                hash_map[pat] = match
            if match not in reverse_map:
                reverse_map[match] = pat
            if pat != reverse_map[match] or hash_map[pat] != match:
                return False
        return True

File: Leetcode/cpp/290.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        std::stringstream ss(s);
        std::string word;
        int index = 0;

        std::unordered_map<char, std::string> char2word;
        std::unordered_map<std::string, char> word2char;
        while (ss >> word) {
            if (index == pattern.length()) {
                return false;
            }
            char chr = pattern[index++];
            if (!char2word.contains(chr) && !word2char.contains(word)) {
                char2word[chr] = word;
                word2char[word] = chr;
            } else if ((!char2word.contains(chr) && word2char.contains(word)) || (char2word.contains(chr) && !word2char.contains(word)) || (char2word[chr] != word || word2char[word] != chr)) {
                return false;
            }
        }
        return index == pattern.length();
    }
};
